src/v/cloud_storage/cache_service.cc

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_storage/logger.h"
+#include "random/generators.h"
+#include "utils/gate_guard.h"
+#include "vlog.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/file.hh>
+#include <seastar/core/fstream.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/seastar.hh>
+#include <seastar/core/smp.hh>
+#include <seastar/util/defer.hh>
+
+#include <cloud_storage/cache_service.h>
+
+#include <cstdint>
+#include <exception>
+#include <optional>
+#include <string>
+
+namespace cloud_storage {
+
+cache::cache(
+  std::filesystem::path cache_dir,
+  size_t max_cache_size,
+  ss::lowres_clock::duration check_period) noexcept
+  : _cache_dir(std::move(cache_dir))
+  , _max_cache_size(max_cache_size)
+  , _check_period(check_period)
+  , _cnt(0) {}
+
+ss::future<uint64_t> cache::get_curr_cache_size() {
+    // todo: implement this
+    return seastar::make_ready_future<uint64_t>(0);
+}
+
+ss::future<> cache::clean_up_cache() {
+    uint64_t curr_cache_size = co_await get_curr_cache_size();
+    if (curr_cache_size >= _max_cache_size) {
+        auto to_delete = curr_cache_size - (uint64_t)(0.8 * _max_cache_size);
+        co_await _walker.walk(_cache_dir.native());
+        auto candidates_for_deletion = _walker.files;
+        uint64_t deleted = 0;
+        while (deleted < to_delete) {
+            auto next_to_delete = candidates_for_deletion.front();
+            candidates_for_deletion.pop_front();
+            deleted += next_to_delete.size;
+            auto unused = invalidate(
```

Comment:
`invalidate` takes a key as a parameter, it adds it to cache directory dir
but paths in the `_walker.files` are absolute paths, so instead of invalidating the key you need to just remove the file
also, it will be useful if you will add vassert that checks that the file is withing the cache directory (just in case if we're walked outside the cache directory)
recoursive_directory_walker should probably also check this just in case there is a bug in the code somewhere

Reply:
done not invalidating, but deleting file. `vasserts` added.